kv/kvserver: TestTxnClearRangeIntents failed

kv/kvserver.TestTxnClearRangeIntents [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2912185&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2912185&tab=artifacts#/) on master @ [ee8b38ef3e0f7a41ae277c127bb9db5989ef47d0](https://github.com/cockroachdb/cockroach/commits/ee8b38ef3e0f7a41ae277c127bb9db5989ef47d0):


```
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:294 +0x3a5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).CheckAllowed(0xc002674d20, 0x0, 0xc0025fe700, 0x8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:225 +0xba
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*EngineIterator).SeekEngineKeyGE(0xc000f24c60, 0xc0025fe700, 0x8, 0x8, 0x0, 0x0, 0x0, 0x7, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/batch.go:256 +0x1a8
        	            	github.com/cockroachdb/cockroach/pkg/storage.ScanSeparatedIntents(0x86209c0, 0xc001f301b0, 0xc0034b2872, 0x2, 0x2, 0xc0034b2874, 0x1, 0x1, 0x3e8, 0xf4240, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:888 +0x385
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.ClearRange(0x85b3400, 0xc00328ec30, 0x7fe871f78c70, 0xc001f301b0, 0x8715ae0, 0xc002ad2540, 0x16783a9e808718ad, 0x0, 0x100000001, 0x1, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_clear_range.go:96 +0x30a
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand(0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0x0, 0x7fe871f78c70, 0xc001f301b0, 0x8715ae0, 0xc002ad2540, 0xc000040280, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:506 +0x319
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch(0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0x7fe871f78c70, 0xc001f301b0, 0x8715ae0, 0xc002ad2540, 0xc000040280, 0xc001447360, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:271 +0x4bd
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWrapper(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0x8715ae0, 0xc002ad2540, 0xc000040280, 0xc001447360, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:657 +0x19d
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWithServersideRefreshes(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0x8715ae0, 0xc002ad2540, 0xc000040280, 0xc001447360, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:625 +0x1e5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatch(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0xc001447360, 0x0, 0x0, 0xc002674d20, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:440 +0x276
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateProposal(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0xc001447360, 0x0, 0x0, 0xc002674d20, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:775 +0x177
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).requestToProposal(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc0025fe6d8, 0x8, 0xc001447360, 0x0, 0x0, 0xc0034b2b70, 0x100000001, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:906 +0xcc
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc001447360, 0xc002ad20e0, 0x0, 0x0, 0xc0034b2b70, 0x100000001, 0x1, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:84 +0x20e
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc001447360, 0xc002ad20e0, 0x0, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:144 +0x8a5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc002b82400, 0x85b3400, 0xc00328ec30, 0xc001447360, 0x70ef010, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:275 +0x409
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc002b82400, 0x85b3400, 0xc002a7eb10, 0x5, 0xc001447360, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:95 +0x629
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:34
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc0006d0800, 0x85b3380, 0xc00019c010, 0x16783a9e808718ad, 0x0, 0x100000001, 0x1, 0x0, 0x5, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x573
        	            	github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x85b3380, 0xc00019c010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:456 +0x175
        	            	github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc000f24b80, 0x85b3380, 0xc00019c010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:392 +0xcb
        	            	github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x85b3380, 0xc00019c010, 0x851f7c0, 0xc000f24b80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:432 +0x195
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestTxnClearRangeIntents(0xc001152180)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/txn_recovery_integration_test.go:398 +0x1b39
        	            	testing.tRunner(0xc001152180, 0x70f0428)
        	            		/usr/local/go/src/testing/testing.go:1123 +0x203
        	            	created by testing.(*T).Run
        	            		/usr/local/go/src/testing/testing.go:1168 +0x5bc
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnClearRangeIntents465701679
--- FAIL: TestTxnClearRangeIntents (0.19s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestTxnClearRangeIntents PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @erikgrinaker
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnClearRangeIntents.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTxnClearRangeIntents [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2913343&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2913343&tab=artifacts#/) on master @ [97955d46cef43897f0281b336e521bc89a5fa400](https://github.com/cockroachdb/cockroach/commits/97955d46cef43897f0281b336e521bc89a5fa400):


```
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:294 +0x3a5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).CheckAllowed(0xc00108daa0, 0x0, 0xc0014ff0e8, 0x8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:225 +0xba
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*EngineIterator).SeekEngineKeyGE(0xc0011e5300, 0xc0014ff0e8, 0x8, 0x8, 0x0, 0x0, 0x0, 0x7, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/batch.go:256 +0x1a8
        	            	github.com/cockroachdb/cockroach/pkg/storage.ScanSeparatedIntents(0x8621fa0, 0xc002dcb7a0, 0xc002eda142, 0x2, 0x2, 0xc002eda144, 0x1, 0x1, 0x3e8, 0xf4240, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:888 +0x385
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.ClearRange(0x85b49e0, 0xc00303df80, 0x7ff85e21e378, 0xc002dcb7a0, 0x87170c0, 0xc002294cb0, 0x167845504ce20b35, 0x0, 0x100000001, 0x1, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_clear_range.go:96 +0x30a
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand(0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0x0, 0x7ff85e21e378, 0xc002dcb7a0, 0x87170c0, 0xc002294cb0, 0xc002538300, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:506 +0x319
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch(0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0x7ff85e21e378, 0xc002dcb7a0, 0x87170c0, 0xc002294cb0, 0xc002538300, 0xc00067fd60, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:271 +0x4bd
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWrapper(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0x87170c0, 0xc002294cb0, 0xc002538300, 0xc00067fd60, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:657 +0x19d
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWithServersideRefreshes(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0x87170c0, 0xc002294cb0, 0xc002538300, 0xc00067fd60, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:625 +0x1e5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatch(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0xc00067fd60, 0x0, 0x0, 0xc00108daa0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:440 +0x276
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateProposal(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0xc00067fd60, 0x0, 0x0, 0xc00108daa0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:782 +0x177
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).requestToProposal(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc0014ff0e0, 0x8, 0xc00067fd60, 0x0, 0x0, 0xc002be6740, 0x100000001, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:913 +0xcc
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc00067fd60, 0xc002294c40, 0x0, 0x0, 0xc002be6740, 0x100000001, 0x1, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:84 +0x20e
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc00067fd60, 0xc002294c40, 0x0, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:144 +0x8a5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc00206f200, 0x85b49e0, 0xc00303df80, 0xc00067fd60, 0x70f0178, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:275 +0x409
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc00206f200, 0x85b49e0, 0xc000217890, 0x5, 0xc00067fd60, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:95 +0x629
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:34
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc0010fd800, 0x85b4960, 0xc000074140, 0x167845504ce20b35, 0x0, 0x100000001, 0x1, 0x0, 0x5, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x573
        	            	github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x85b4960, 0xc000074140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:456 +0x175
        	            	github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc0011e5240, 0x85b4960, 0xc000074140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:392 +0xcb
        	            	github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x85b4960, 0xc000074140, 0x8520d60, 0xc0011e5240, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:432 +0x195
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestTxnClearRangeIntents(0xc00143c180)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/txn_recovery_integration_test.go:398 +0x1b39
        	            	testing.tRunner(0xc00143c180, 0x70f1590)
        	            		/usr/local/go/src/testing/testing.go:1123 +0x203
        	            	created by testing.(*T).Run
        	            		/usr/local/go/src/testing/testing.go:1168 +0x5bc
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnClearRangeIntents846422537
--- FAIL: TestTxnClearRangeIntents (0.20s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestTxnClearRangeIntents PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @erikgrinaker
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnClearRangeIntents.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTxnClearRangeIntents [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2913557&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2913557&tab=artifacts#/) on master @ [ea6dc019d083434e0df6b3181efc7522405f79b6](https://github.com/cockroachdb/cockroach/commits/ea6dc019d083434e0df6b3181efc7522405f79b6):


```
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:294 +0x3a5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).CheckAllowed(0xc002cc4d80, 0x0, 0xc002bfc4d0, 0x8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:225 +0xba
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*EngineIterator).SeekEngineKeyGE(0xc002466860, 0xc002bfc4d0, 0x8, 0x8, 0x0, 0x0, 0x0, 0x7, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/batch.go:256 +0x1a8
        	            	github.com/cockroachdb/cockroach/pkg/storage.ScanSeparatedIntents(0x8621fa0, 0xc002834360, 0xc001d34a62, 0x2, 0x2, 0xc001d34a64, 0x1, 0x1, 0x3e8, 0xf4240, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:888 +0x385
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.ClearRange(0x85b49e0, 0xc003611fb0, 0x7ff80d3902f0, 0xc002834360, 0x87170c0, 0xc002b405b0, 0x167847c575a7b3da, 0x0, 0x100000001, 0x1, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_clear_range.go:96 +0x30a
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand(0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0x0, 0x7ff80d3902f0, 0xc002834360, 0x87170c0, 0xc002b405b0, 0xc000022380, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:506 +0x319
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch(0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0x7ff80d3902f0, 0xc002834360, 0x87170c0, 0xc002b405b0, 0xc000022380, 0xc001719720, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:271 +0x4bd
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWrapper(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0x87170c0, 0xc002b405b0, 0xc000022380, 0xc001719720, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:657 +0x19d
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWithServersideRefreshes(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0x87170c0, 0xc002b405b0, 0xc000022380, 0xc001719720, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:625 +0x1e5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatch(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0xc001719720, 0x0, 0x0, 0xc002cc4d80, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:440 +0x276
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateProposal(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0xc001719720, 0x0, 0x0, 0xc002cc4d80, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:782 +0x177
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).requestToProposal(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc002bfc4b8, 0x8, 0xc001719720, 0x0, 0x0, 0xc001d34da0, 0x100000001, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:913 +0xcc
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc001719720, 0xc002b400e0, 0x0, 0x0, 0xc001d34da0, 0x100000001, 0x1, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:84 +0x20e
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc001719720, 0xc002b400e0, 0x0, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:144 +0x8a5
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc002047200, 0x85b49e0, 0xc003611fb0, 0xc001719720, 0x70f0178, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:275 +0x409
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc002047200, 0x85b49e0, 0xc00277ddd0, 0x5, 0xc001719720, 0x0, 0x0)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:95 +0x629
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:34
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc001245800, 0x85b4960, 0xc000126010, 0x167847c575a7b3da, 0x0, 0x100000001, 0x1, 0x0, 0x5, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x573
        	            	github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x85b4960, 0xc000126010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:456 +0x175
        	            	github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc002466760, 0x85b4960, 0xc000126010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:392 +0xcb
        	            	github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x85b4960, 0xc000126010, 0x8520d60, 0xc002466760, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:432 +0x195
        	            	github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestTxnClearRangeIntents(0xc0027e2180)
        	            		/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/txn_recovery_integration_test.go:398 +0x1b39
        	            	testing.tRunner(0xc0027e2180, 0x70f1590)
        	            		/usr/local/go/src/testing/testing.go:1123 +0x203
        	            	created by testing.(*T).Run
        	            		/usr/local/go/src/testing/testing.go:1168 +0x5bc
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnClearRangeIntents433252767
--- FAIL: TestTxnClearRangeIntents (0.19s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestTxnClearRangeIntents PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @erikgrinaker
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnClearRangeIntents.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/concurrency/concurrency_manager.go
+++ b/pkg/kv/kvserver/concurrency/concurrency_manager.go
@@ -465,6 +465,11 @@ func (g *Guard) LatchSpans() *spanset.SpanSet {
 	return g.Req.LatchSpans
 }
 
+// LockSpans returns the maximal set of lock spans that the request will access.
+func (g *Guard) LockSpans() *spanset.SpanSet {
+	return g.Req.LockSpans
+}
+
 // HoldingLatches returned whether the guard is holding latches or not.
 func (g *Guard) HoldingLatches() bool {
 	return g != nil && g.lg != nil

--- a/pkg/kv/kvserver/replica_proposal.go
+++ b/pkg/kv/kvserver/replica_proposal.go
@@ -765,7 +765,7 @@ func (r *Replica) evaluateProposal(
 	idKey kvserverbase.CmdIDKey,
 	ba *roachpb.BatchRequest,
 	lul hlc.Timestamp,
-	latchSpans *spanset.SpanSet,
+	latchSpans, lockSpans *spanset.SpanSet,
 ) (*result.Result, bool, *roachpb.Error) {
 	if ba.Timestamp.IsEmpty() {
 		return nil, false, roachpb.NewErrorf("can't propose Raft command with zero timestamp")
@@ -781,7 +781,7 @@ func (r *Replica) evaluateProposal(
 	//
 	// TODO(tschottdorf): absorb all returned values in `res` below this point
 	// in the call stack as well.
-	batch, ms, br, res, pErr := r.evaluateWriteBatch(ctx, idKey, ba, lul, latchSpans)
+	batch, ms, br, res, pErr := r.evaluateWriteBatch(ctx, idKey, ba, lul, latchSpans, lockSpans)
 
 	// Note: reusing the proposer's batch when applying the command on the
 	// proposer was explored as an optimization but resulted in no performance
@@ -912,9 +912,9 @@ func (r *Replica) requestToProposal(
 	ba *roachpb.BatchRequest,
 	st kvserverpb.LeaseStatus,
 	lul hlc.Timestamp,
-	latchSpans *spanset.SpanSet,
+	latchSpans, lockSpans *spanset.SpanSet,
 ) (*ProposalData, *roachpb.Error) {
-	res, needConsensus, pErr := r.evaluateProposal(ctx, idKey, ba, lul, latchSpans)
+	res, needConsensus, pErr := r.evaluateProposal(ctx, idKey, ba, lul, latchSpans, lockSpans)
 
 	// Fill out the results even if pErr != nil; we'll return the error below.
 	proposal := &ProposalData{

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -86,7 +86,7 @@ func (r *Replica) evalAndPropose(
 ) (chan proposalResult, func(), int64, *roachpb.Error) {
 	defer tok.DoneIfNotMoved(ctx)
 	idKey := makeIDKey()
-	proposal, pErr := r.requestToProposal(ctx, idKey, ba, st, lul, g.LatchSpans())
+	proposal, pErr := r.requestToProposal(ctx, idKey, ba, st, lul, g.LatchSpans(), g.LockSpans())
 	log.Event(proposal.ctx, "evaluated request")
 
 	// If the request hit a server-side concurrency retry error, immediately

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -8115,7 +8115,7 @@ func TestReplicaRefreshPendingCommandsTicks(t *testing.T) {
 		ba.Timestamp = tc.Clock().Now()
 		ba.Add(&roachpb.PutRequest{RequestHeader: roachpb.RequestHeader{Key: roachpb.Key(id)}})
 		st := r.CurrentLeaseStatus(ctx)
-		cmd, pErr := r.requestToProposal(ctx, kvserverbase.CmdIDKey(id), &ba, st, hlc.Timestamp{}, &allSpans)
+		cmd, pErr := r.requestToProposal(ctx, kvserverbase.CmdIDKey(id), &ba, st, hlc.Timestamp{}, &allSpans, &allSpans)
 		if pErr != nil {
 			t.Fatal(pErr)
 		}
@@ -8237,7 +8237,7 @@ func TestReplicaRefreshMultiple(t *testing.T) {
 
 	incCmdID = makeIDKey()
 	atomic.StoreInt32(&filterActive, 1)
-	proposal, pErr := repl.requestToProposal(ctx, incCmdID, &ba, repl.CurrentLeaseStatus(ctx), hlc.Timestamp{}, &allSpans)
+	proposal, pErr := repl.requestToProposal(ctx, incCmdID, &ba, repl.CurrentLeaseStatus(ctx), hlc.Timestamp{}, &allSpans, &allSpans)
 	if pErr != nil {
 		t.Fatal(pErr)
 	}
@@ -8687,7 +8687,7 @@ func TestReplicaEvaluationNotTxnMutation(t *testing.T) {
 	assignSeqNumsForReqs(txn, &txnPut, &txnPut2)
 	origTxn := txn.Clone()
 
-	batch, _, _, _, pErr := tc.repl.evaluateWriteBatch(ctx, makeIDKey(), &ba, hlc.Timestamp{}, &allSpans)
+	batch, _, _, _, pErr := tc.repl.evaluateWriteBatch(ctx, makeIDKey(), &ba, hlc.Timestamp{}, &allSpans, &allSpans)
 	defer batch.Close()
 	if pErr != nil {
 		t.Fatal(pErr)
@@ -12842,7 +12842,7 @@ func TestContainsEstimatesClampProposal(t *testing.T) {
 		ba.Timestamp = tc.Clock().Now()
 		req := putArgs(roachpb.Key("some-key"), []byte("some-value"))
 		ba.Add(&req)
-		proposal, err := tc.repl.requestToProposal(ctx, cmdIDKey, &ba, tc.repl.CurrentLeaseStatus(ctx), hlc.Timestamp{}, &allSpans)
+		proposal, err := tc.repl.requestToProposal(ctx, cmdIDKey, &ba, tc.repl.CurrentLeaseStatus(ctx), hlc.Timestamp{}, &allSpans, &allSpans)
 		if err != nil {
 			t.Error(err)
 		}

--- a/pkg/kv/kvserver/replica_write.go
+++ b/pkg/kv/kvserver/replica_write.go
@@ -15,6 +15,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/result"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/ctpb"
@@ -420,7 +421,7 @@ func (r *Replica) evaluateWriteBatch(
 	idKey kvserverbase.CmdIDKey,
 	ba *roachpb.BatchRequest,
 	lul hlc.Timestamp,
-	latchSpans *spanset.SpanSet,
+	latchSpans, lockSpans *spanset.SpanSet,
 ) (storage.Batch, enginepb.MVCCStats, *roachpb.BatchResponse, result.Result, *roachpb.Error) {
 	log.Event(ctx, "executing read-write batch")
 
@@ -436,7 +437,7 @@ func (r *Replica) evaluateWriteBatch(
 		return nil, enginepb.MVCCStats{}, nil, result.Result{}, pErr
 	}
 	if ok {
-		res := r.evaluate1PC(ctx, idKey, ba, latchSpans)
+		res := r.evaluate1PC(ctx, idKey, ba, latchSpans, lockSpans)
 		switch res.success {
 		case onePCSucceeded:
 			return res.batch, res.stats, res.br, res.res, nil
@@ -452,7 +453,7 @@ func (r *Replica) evaluateWriteBatch(
 	ms := new(enginepb.MVCCStats)
 	rec := NewReplicaEvalContext(r, latchSpans)
 	batch, br, res, pErr := r.evaluateWriteBatchWithServersideRefreshes(
-		ctx, idKey, rec, ms, ba, lul, latchSpans, nil /* deadline */)
+		ctx, idKey, rec, ms, ba, lul, latchSpans, lockSpans, nil /* deadline */)
 	return batch, *ms, br, res, pErr
 }
 
@@ -493,7 +494,7 @@ func (r *Replica) evaluate1PC(
 	ctx context.Context,
 	idKey kvserverbase.CmdIDKey,
 	ba *roachpb.BatchRequest,
-	latchSpans *spanset.SpanSet,
+	latchSpans, lockSpans *spanset.SpanSet,
 ) (onePCRes onePCResult) {
 	log.VEventf(ctx, 2, "attempting 1PC execution")
 
@@ -526,10 +527,10 @@ func (r *Replica) evaluate1PC(
 	ms := new(enginepb.MVCCStats)
 	if ba.CanForwardReadTimestamp {
 		batch, br, res, pErr = r.evaluateWriteBatchWithServersideRefreshes(
-			ctx, idKey, rec, ms, &strippedBa, localUncertaintyLimit, latchSpans, etArg.Deadline)
+			ctx, idKey, rec, ms, &strippedBa, localUncertaintyLimit, latchSpans, lockSpans, etArg.Deadline)
 	} else {
 		batch, br, res, pErr = r.evaluateWriteBatchWrapper(
-			ctx, idKey, rec, ms, &strippedBa, localUncertaintyLimit, latchSpans)
+			ctx, idKey, rec, ms, &strippedBa, localUncertaintyLimit, latchSpans, lockSpans)
 	}
 
 	if pErr != nil || (!ba.CanForwardReadTimestamp && ba.Timestamp != br.Timestamp) {
@@ -622,7 +623,7 @@ func (r *Replica) evaluateWriteBatchWithServersideRefreshes(
 	ms *enginepb.MVCCStats,
 	ba *roachpb.BatchRequest,
 	lul hlc.Timestamp,
-	latchSpans *spanset.SpanSet,
+	latchSpans, lockSpans *spanset.SpanSet,
 	deadline *hlc.Timestamp,
 ) (batch storage.Batch, br *roachpb.BatchResponse, res result.Result, pErr *roachpb.Error) {
 	goldenMS := *ms
@@ -636,7 +637,8 @@ func (r *Replica) evaluateWriteBatchWithServersideRefreshes(
 			batch.Close()
 		}
 
-		batch, br, res, pErr = r.evaluateWriteBatchWrapper(ctx, idKey, rec, ms, ba, lul, latchSpans)
+		batch, br, res, pErr = r.evaluateWriteBatchWrapper(
+			ctx, idKey, rec, ms, ba, lul, latchSpans, lockSpans)
 
 		var success bool
 		if pErr == nil {
@@ -665,9 +667,9 @@ func (r *Replica) evaluateWriteBatchWrapper(
 	ms *enginepb.MVCCStats,
 	ba *roachpb.BatchRequest,
 	lul hlc.Timestamp,
-	latchSpans *spanset.SpanSet,
+	latchSpans, lockSpans *spanset.SpanSet,
 ) (storage.Batch, *roachpb.BatchResponse, result.Result, *roachpb.Error) {
-	batch, opLogger := r.newBatchedEngine(latchSpans)
+	batch, opLogger := r.newBatchedEngine(latchSpans, lockSpans)
 	br, res, pErr := evaluateBatch(ctx, idKey, batch, rec, ms, ba, lul, false /* readOnly */)
 	if pErr == nil {
 		if opLogger != nil {
@@ -683,7 +685,9 @@ func (r *Replica) evaluateWriteBatchWrapper(
 // are enabled, it also returns an engine.OpLoggerBatch. If non-nil, then this
 // OpLogger is attached to the returned engine.Batch, recording all operations.
 // Its recording should be attached to the Result of request evaluation.
-func (r *Replica) newBatchedEngine(spans *spanset.SpanSet) (storage.Batch, *storage.OpLoggerBatch) {
+func (r *Replica) newBatchedEngine(
+	latchSpans, lockSpans *spanset.SpanSet,
+) (storage.Batch, *storage.OpLoggerBatch) {
 	batch := r.store.Engine().NewBatch()
 	if !batch.ConsistentIterators() {
 		// This is not currently needed for correctness, but future optimizations
@@ -728,6 +732,17 @@ func (r *Replica) newBatchedEngine(spans *spanset.SpanSet) (storage.Batch, *stor
 		batch = opLogger
 	}
 	if util.RaceEnabled {
+		// To account for separated intent accesses, we translate the lock spans
+		// to lock table spans.
+		spans := latchSpans.Copy()
+		lockSpans.Iterate(func(sa spanset.SpanAccess, _ spanset.SpanScope, span spanset.Span) {
+			ltKey, _ := keys.LockTableSingleKey(span.Key, nil)
+			var ltEndKey roachpb.Key
+			if span.EndKey != nil {
+				ltEndKey, _ = keys.LockTableSingleKey(span.EndKey, nil)
+			}
+			spans.AddNonMVCC(sa, roachpb.Span{Key: ltKey, EndKey: ltEndKey})
+		})
 		// During writes we may encounter a versioned value newer than the request
 		// timestamp, and may have to retry at a higher timestamp. This is still
 		// safe as we're only ever writing at timestamps higher than the timestamp

--- a/pkg/kv/kvserver/spanset/spanset.go
+++ b/pkg/kv/kvserver/spanset/spanset.go
@@ -114,6 +114,31 @@ func (s *SpanSet) Empty() bool {
 	return s.Len() == 0
 }
 
+// Copy copies the SpanSet.
+func (s *SpanSet) Copy() *SpanSet {
+	n := &SpanSet{}
+	for sa := SpanAccess(0); sa < NumSpanAccess; sa++ {
+		for ss := SpanScope(0); ss < NumSpanScope; ss++ {
+			n.spans[sa][ss] = append(n.spans[sa][ss], s.spans[sa][ss]...)
+		}
+	}
+	return n
+}
+
+// Iterate iterates over a SpanSet, calling the given function.
+func (s *SpanSet) Iterate(f func(SpanAccess, SpanScope, Span)) {
+	if s == nil {
+		return
+	}
+	for sa := SpanAccess(0); sa < NumSpanAccess; sa++ {
+		for ss := SpanScope(0); ss < NumSpanScope; ss++ {
+			for _, span := range s.spans[sa][ss] {
+				f(sa, ss, span)
+			}
+		}
+	}
+}
+
 // Reserve space for N additional spans.
 func (s *SpanSet) Reserve(access SpanAccess, scope SpanScope, n int) {
 	existing := s.spans[access][scope]

--- a/pkg/kv/kvserver/spanset/spanset_test.go
+++ b/pkg/kv/kvserver/spanset/spanset_test.go
@@ -49,6 +49,54 @@ func TestSpanSetGetSpansScope(t *testing.T) {
 	}
 }
 
+func TestSpanSetCopy(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ss := new(SpanSet)
+	ss.AddMVCC(SpanReadOnly, roachpb.Span{Key: roachpb.Key("abc")}, hlc.Timestamp{WallTime: 123, Logical: 7})
+	ss.AddNonMVCC(SpanReadWrite, roachpb.Span{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")})
+
+	c := ss.Copy()
+	require.Equal(t, ss, c)
+
+	// modifying element in ss should not modify copy
+	ss.AddNonMVCC(SpanReadOnly, roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")})
+	require.NotEqual(t, ss, c)
+}
+
+func TestSpanSetIterate(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	spA := roachpb.Span{Key: roachpb.Key("a")}
+	spRO := roachpb.Span{Key: roachpb.Key("r"), EndKey: roachpb.Key("o")}
+	spRW := roachpb.Span{Key: roachpb.Key("r"), EndKey: roachpb.Key("w")}
+	spLocal := roachpb.Span{Key: keys.RangeLastGCKey(1)}
+
+	ss := new(SpanSet)
+	ss.AddNonMVCC(SpanReadOnly, spLocal)
+	ss.AddNonMVCC(SpanReadOnly, spRO)
+	ss.AddNonMVCC(SpanReadOnly, spA)
+	ss.AddNonMVCC(SpanReadWrite, spRW)
+
+	type item struct {
+		sa   SpanAccess
+		ss   SpanScope
+		span Span
+	}
+	expect := []item{
+		{sa: SpanReadOnly, ss: SpanGlobal, span: Span{Span: spRO}},
+		{sa: SpanReadOnly, ss: SpanGlobal, span: Span{Span: spA}},
+		{sa: SpanReadOnly, ss: SpanLocal, span: Span{Span: spLocal}},
+		{sa: SpanReadWrite, ss: SpanGlobal, span: Span{Span: spRW}},
+	}
+	items := []item{}
+	ss.Iterate(func(sa SpanAccess, ss SpanScope, span Span) {
+		items = append(items, item{sa: sa, ss: ss, span: span})
+	})
+
+	require.Equal(t, expect, items)
+}
+
 func TestSpanSetMerge(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

--- a/pkg/kv/kvserver/txn_recovery_integration_test.go
+++ b/pkg/kv/kvserver/txn_recovery_integration_test.go
@@ -20,7 +20,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -340,7 +339,6 @@ func TestTxnRecoveryFromStagingWithHighPriority(t *testing.T) {
 // are disabled. See also: https://github.com/cockroachdb/cockroach/issues/46764
 func TestTxnClearRangeIntents(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.UnderRaceWithIssue(t, 64088, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()

